homeassistant/components/tts/__init__.py

```diff
@@ -0,0 +1,143 @@
+            language = call.data.get(ATTR_LANGUAGE, config.get(ATTR_LANGUAGE,
+                                                               "en-us"))
+            rate = call.data.get(ATTR_RATE, config.get(ATTR_RATE, 0))
+            codec = call.data.get(ATTR_CODEC, config.get(ATTR_CODEC, "MP3"))
+            audio_format = call.data.get(ATTR_FORMAT,
+                                         config.get(ATTR_FORMAT,
+                                                    "44khz_16bit_stereo"))
+
+            hashed_text = hashlib.sha1(text.encode('utf-8')).hexdigest()
+
+            full_filename = "{}-{}-{}.{}".format(hashed_text, language,
+                                                 platform_name,
+                                                 codec.lower())
+
+            file_path = os.path.join(tts_dir, full_filename)
+
+            if text is None:
+                _LOGGER.error(
+                    'Received call to %s without attribute %s',
+                    call.service, ATTR_TEXT)
+                return
+
+            if (os.path.isfile(file_path)) is False or \
+               (allow_cached_file is False):
+                try:
+                    tts_service.speak(file_path, text, language, rate, codec,
+                                      audio_format)
+                # pylint: disable=bare-except
+                except:
+                    _LOGGER.error('Error trying to speak using %s',
+                                  platform_name)
+                    return
+
+            if play_file is True or entity_id is None:
+                content_url = "{}/api/tts/{}".format(hass.config.api.base_url,
+                                                     full_filename)
+                hass.services.call(core.DOMAIN, SERVICE_PLAY_MEDIA, {
+                    ATTR_ENTITY_ID: entity_id,
+                    ATTR_MEDIA_CONTENT_ID: content_url,
+                    ATTR_MEDIA_CONTENT_TYPE: MEDIA_TYPE_MUSIC
+                })
+
+        service_tts = p_config.get(CONF_NAME, platform)
+        service_call_handler = partial(generate_speech, tts_engine,
+                                       service_tts)
+        hass.services.register(DOMAIN, service_tts, service_call_handler,
+                               descriptions.get(SERVICE_TTS))
+        success = True
+
+    hass.http.register_path(
+        'GET', re.compile(r'/api/tts/(?P<tts_filename>.*)'),
+        _handle_get_tts_file)
+
+    return success
+
+
+def _handle_get_tts_file(handler, path_match, data):
+    """Return a TTS file."""
+    handler.write_file(os.path.join(handler.server.hass.config.path(DOMAIN),
+                                    path_match.group('tts_filename')))
+
+# pylint: disable=too-few-public-methods
+
+
+class BaseTTSService(object):
+    """An abstract class for TTS services."""
+
+    # pylint: disable=too-many-arguments
+    def speak(self, file_path, text, language, rate, codec, audio_format):
```

Comment:
Let's make all but file_path and text optional arguments. That way we can easily extend it in the future without having to update all platforms.

I also think that the method name `speak` is not descriptive of what this method does. Something like `get_speech` sounds better I think